src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -266,25 +277,44 @@ internal static SafeSslHandle AllocateSslHandle(SafeFreeSslCredentials credentia
             SafeSslContextHandle? newCtxHandle = null;
             SslProtocols protocols = CalculateEffectiveProtocols(sslAuthenticationOptions);
             bool hasAlpn = sslAuthenticationOptions.ApplicationProtocols != null && sslAuthenticationOptions.ApplicationProtocols.Count != 0;
-            bool cacheSslContext = !DisableTlsResume && sslAuthenticationOptions.EncryptionPolicy == EncryptionPolicy.RequireEncryption &&
-                    sslAuthenticationOptions.IsServer &&
-                    sslAuthenticationOptions.CertificateContext != null &&
-                    sslAuthenticationOptions.CertificateContext.SslContexts != null &&
-                    sslAuthenticationOptions.CipherSuitesPolicy == null;
+            bool cacheSslContext = !DisableTlsResume && sslAuthenticationOptions.EncryptionPolicy == EncryptionPolicy.RequireEncryption && sslAuthenticationOptions.CipherSuitesPolicy == null &&
+                     ((sslAuthenticationOptions.IsServer &&
+                       sslAuthenticationOptions.CertificateContext != null &&
+                       sslAuthenticationOptions.CertificateContext.SslContexts != null) ||
+                      (sslAuthenticationOptions.IsClient &&
+                      // since SNI us our key, we don't wont to resume sessions without it.
+                      !string.IsNullOrEmpty(sslAuthenticationOptions.TargetHost) &&
+                       // on client avoid resume with certificates
+                       sslAuthenticationOptions.CertificateContext == null &&
+                       sslAuthenticationOptions.CertSelectionDelegate == null));
```

Comment:
I'd probably break this up to avoid mixing ands and ors in such a long expression.

```C#
bool cacheSslContext = everything && before && IsClient/IsServer;

if (cacheSslContext && sslAuthenticationOptions.IsServer)
{
    if (sslAuthenticationOptions.CertificateContext?.SslContexts == null)
    {
        cacheSslContext = false;
    }
...
```

Reply:
I like the first suggestion. updated.